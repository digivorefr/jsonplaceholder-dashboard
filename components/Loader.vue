<template>
  <div class="loader" />
</template>

<script lang="
    ts"
  >
import Vue from 'vue'

export default Vue.extend({
  name: 'ComponentLoader'
})
</script>
  </div>
</template>
