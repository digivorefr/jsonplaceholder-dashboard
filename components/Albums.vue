<template>
  <div :class="`user__albums${($fetchState.error !== null) ? ' error' : ''}`">
    <h2>Albums</h2>
    <h3 v-if="$fetchState.error !== null">
      An error occured.
    </h3>
    <Loader v-else-if="$fetchState.pending === true" />
    <h3 v-else-if="albums.length === 0">
      No albums
    </h3>
    <div
      v-else
      class="user__albums__wrapper"
    >
      <div class="user__albums__grid">
        <div
          v-for="album in albums"
          :key="album.id"
          class="user__album"
        >
          <h4 class="user__album__title">
            {{ album.title }}
          </h4>
          <div
            v-if="album.pictures.length > 0"
            class="user__album__gallery"
          >
            <div
              v-for="picture in album.pictures"
              :key="picture.id"
              class="user__album__gallery__picture"
            >
              <Loader />
              <img
                :src="picture.thumbnailUrl"
                loading="lazy"
              >
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import Loader from '~/components/Loader.vue'

export type Picture = {
  albumId: number;
  id: number;
  title: string;
  url: string;
  thumbnailUrl: string;
}

export type Album = {
  userId: number;
  id: number;
  title: string;
  pictures: Picture[],
}

export default Vue.extend({
  name: 'UserTodos',
  components: { Loader },
  props: {
    id: {
      type: Number,
      required: true
    }
  },
  data() {
    return {
      albums: [] as Album[]
    }
  },
  async fetch() {
    this.albums = await this.fetchAlbums()
  },
  methods: {
    async fetchAlbums() {
      const albums = await this.$axios.$get(`/albums?userId=${this.id}`) as Album[]
      const picturePromises = albums.map(album => this.$axios.$get(`/albums/${album.id}/photos`))
      const pictures = await Promise.all(picturePromises)
      return albums.map((album, index) => ({ ...album, pictures: pictures[index] }))
    }
  }
})
</script>
