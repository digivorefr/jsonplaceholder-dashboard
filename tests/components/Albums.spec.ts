import { mount, Wrapper } from '@vue/test-utils'
import Albums from '@/components/Albums.vue'
import Loader from '@/components/Loader.vue'

const albumsMock = [
  {
    userId: 1,
    id: 1,
    title: 'quidem molestiae enim'
  },
  {
    userId: 1,
    id: 2,
    title: 'sunt qui excepturi placeat culpa'
  }
]

const photosMock = [
  {
    albumId: 1,
    id: 1,
    title: 'accusamus beatae ad facilis cum similique qui sunt',
    url: 'https://via.placeholder.com/600/92c952',
    thumbnailUrl: 'https://via.placeholder.com/150/92c952'
  },
  {
    albumId: 1,
    id: 2,
    title: 'reprehenderit est deserunt velit ipsam',
    url: 'https://via.placeholder.com/600/771796',
    thumbnailUrl: 'https://via.placeholder.com/150/771796'
  }
]

const axiosMock = {
  $get: (url: string) => {
    if (url.match(/albums\?userId=[\d]*/)) {
      return albumsMock
    }
    if (url.match(/albums\/[\d]*\/photos/)) {
      return photosMock
    }
    return []
  }
}

describe('Albums', () => {
  test('renders correctly', () => {
    const wrapper = mount(Albums, {
      mocks: {
        $fetchState: { pending: true, error: true, timestamp: Date.now() }
      }
    })
    expect(wrapper.vm).toBeTruthy()
    expect(wrapper.exists()).toBe(true)
    wrapper.destroy()
  })
  test('show error if fetch fails', () => {
    const wrapper = mount(Albums, {
      mocks: {
        $fetchState: { pending: null, error: true, timestamp: Date.now() }
      }
    })
    expect(wrapper.html()).toContain('An error occured.')
    wrapper.destroy()
  })
  test('show loader while fetching', () => {
    const wrapper = mount(Albums, {
      mocks: {
        $fetchState: { pending: true, error: null, timestamp: Date.now() }
      }
    })
    expect(wrapper.findComponent(Loader))
    wrapper.destroy()
  })
  test('show no albums text', () => {
    const wrapper = mount(Albums, {
      data() {
        return { albums: [] }
      },
      mocks: {
        $fetchState: { pending: null, error: null, timestamp: Date.now() }
      }
    })
    expect(wrapper.html()).toContain('No albums')
    wrapper.destroy()
  })
  test('fetch and display albums', async () => {
    const wrapper: Wrapper<Vue & { [key: string]: any }> = mount(Albums, {
      mocks: {
        $fetchState: { pending: null, error: null, timestamp: Date.now() },
        $axios: axiosMock
      }
    })
    wrapper.vm.$data.albums = await wrapper.vm.fetchAlbums()
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.$data.albums.length).toBe(2)
    expect(wrapper.vm.$data.albums[0].pictures.length).toBe(2)
    expect(wrapper.findAll('.user__album').length).toBe(2)
    expect(wrapper.element).toMatchSnapshot()

    wrapper.destroy()
  })
})
