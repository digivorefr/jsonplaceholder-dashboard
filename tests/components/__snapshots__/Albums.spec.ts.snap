// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Albums fetch and display albums 1`] = `
<div
  class="user__albums"
>
  <h2>
    Albums
  </h2>
   
  <div
    class="user__albums__wrapper"
  >
    <div
      class="user__albums__grid"
    >
      <div
        class="user__album"
      >
        <h4
          class="user__album__title"
        >
          
          quidem molestiae enim
        
        </h4>
         
        <div
          class="user__album__gallery"
        >
          <div
            class="user__album__gallery__picture"
          >
            <div
              class="loader"
            />
             
            <img
              loading="lazy"
              src="https://via.placeholder.com/150/92c952"
            />
          </div>
          <div
            class="user__album__gallery__picture"
          >
            <div
              class="loader"
            />
             
            <img
              loading="lazy"
              src="https://via.placeholder.com/150/771796"
            />
          </div>
        </div>
      </div>
      <div
        class="user__album"
      >
        <h4
          class="user__album__title"
        >
          
          sunt qui excepturi placeat culpa
        
        </h4>
         
        <div
          class="user__album__gallery"
        >
          <div
            class="user__album__gallery__picture"
          >
            <div
              class="loader"
            />
             
            <img
              loading="lazy"
              src="https://via.placeholder.com/150/92c952"
            />
          </div>
          <div
            class="user__album__gallery__picture"
          >
            <div
              class="loader"
            />
             
            <img
              loading="lazy"
              src="https://via.placeholder.com/150/771796"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
